core/src/code/utils.rs

```diff
@@ -331,73 +295,74 @@ pub fn check_data_section(module: &Module, check_stack_end: bool) -> Result<(),
     Ok(())
 }
 
-pub fn check_and_canonize_gear_stack_end(module: &mut Module) -> Result<(), CodeError> {
-    let Some(StackEndInfo {
-        offset: stack_end_offset,
-        is_mutable: stack_end_global_is_mutable,
-    }) = get_stack_end_info(module)?
+fn get_stack_end_offset(module: &Module) -> Result<Option<u32>, CodeError> {
+    let Some((export_index, global_index)) =
+        get_export_global_with_index(module, STACK_END_EXPORT_NAME)
     else {
-        return Ok(());
+        return Ok(None);
     };
 
-    // If [STACK_END_EXPORT_NAME] points to mutable global, then make new const global
-    // with the same init expr and change the export internal to point to the new global.
-    if stack_end_global_is_mutable {
-        // Panic is impossible, because we have checked above, that global section exists.
-        let global_section = module
-            .global_section_mut()
-            .unwrap_or_else(|| unreachable!("Cannot find global section"));
-        let new_global_index = u32::try_from(global_section.entries().len())
-            .map_err(|_| StackEndError::GlobalIndexOverflow)?;
-        global_section.entries_mut().push(GlobalEntry::new(
-            GlobalType::new(parity_wasm::elements::ValueType::I32, false),
-            InitExpr::new(vec![
-                Instruction::I32Const(stack_end_offset),
-                Instruction::End,
-            ]),
-        ));
-
-        // Panic is impossible, because we have checked above,
-        // that stack end export exists and it points to global.
-        get_export_global_index_mut(module, STACK_END_EXPORT_NAME)
-            .map(|global_index| *global_index = new_global_index)
-            .unwrap_or_else(|| unreachable!("Cannot find stack end export"))
+    Ok(Some(
+        get_init_expr_const_i32(
+            get_export_global_entry(module, export_index, global_index)?.init_expr(),
+        )
+        .ok_or(StackEndError::Initialization)? as u32,
+    ))
+}
+
+pub fn check_and_canonize_gear_stack_end(
+    module: &mut Module,
+    static_pages: WasmPage,
+) -> Result<Option<WasmPage>, CodeError> {
+    let Some(stack_end_offset) = get_stack_end_offset(module)? else {
+        return Ok(None);
+    };
+
+    // Remove stack end export from module.
+    // Panic below is impossible, because we have checked above, that export section exists.
+    module
```

Comment:
it seems more correct to perform removal after further checks (even if its identical behaviour for now, something may change in future)

Reply:
It's not correct cause further checks also checks that we have no mutable globals exports, so we must to remove it here